### ISSUE TITLE
Rule idea: flag literal attributes in monkeypatching/mocking APIs

### ISSUE THREAD

**Issue by woodruffw** [astral-sh/ruff#22013]
### Summary

This is a rough idea that probably needs refining/discussion 🙂 

## Overview

Pytest and others expose monkeypatch/mocking APIs like:

```python
monkeypatch.setattr(some_module, "my_important_api", replacement)
```

(and same for `delattr`.)

When called, this temporarily monkeypatches `some_module.my_important_api` with `replacement` (which could be an object, callable, etc.).

## Problem statement

The above works, but has a footgun: if the user performs a refactor (e.g. from `my_important_api` to `my_critical_api`), then may miss this monkeypatch (since it's a string literal, not a usage/reference). This is particularly likely to happen with IDE/LSP refactors, since syntax-aware rename-all functionality can't look at these kinds of non-type string literals by design.

The impact of this depends: sometimes users notice it immediately (since their monkeypatch stops working), while other times it makes tests subtly incorrect or unreliable in ways that aren't noticed until much later.

## Solution statement

In general, users should _probably_ write:

```python
monkeypatch.setattr(some_module, some_module.my_important_api.__name__, replacement)
```

That'll behave the same as above, _but_ will ensure that refactors/mechanical renames propagate as expected.

## Problems

1. This won't work with dynamic attributes; not sure if this is a problem or not.
2. Using `__name__` is arguably pretty unsightly and is maybe considered unidiomatic, not sure 🙂 

## Alternatives

1. Do nothing.
2. Maybe Pytest and others with monkeypatching APIs should support functions as first-class "nameable" objects? In other words, Ruff could instead recommend:

    ```python
    monkeypatch.setattr(some_module, some_module.my_important_api.__name__, replacement)
    ```
    
    ...and Pytest would do the `__name__` access internally. This would require upstream changes, but would be a lot visually cleaner from the user's side.

## Other references

xref https://github.com/astral-sh/ruff/issues/12850 for another (but I think unrelated) mocking/monkeypatching feature request.

**Comment by carljm** [comment]
This seems like something that needs to be cross-module and have type information.

**Comment by ntBre** [comment]
If I understand correctly, we could get by without type inference if we're just looking for `monkeypatch.setattr` with a string as the second argument and recommending to use `__name__` instead. I think we'd only need type inference if we were trying to resolve the definition of the string. But I could definitely be missing something.

**Comment by carljm** [comment]
Oh, yeah, I misunderstood the proposal. A type-aware lint rule could actually tell you when your patch is patching something that doesn't exist, but a non-type-aware rule could just tell you to not use strings.

**Comment by woodruffw** [comment]
Yeah, sorry for the confusion! I was thinking of this as a purely naive lint, i.e. no type or actual attribute awareness.

**Comment by dataCenter430** [comment]
Hi, @woodruffw thanks for reporting this.
I'd love to work on this as a contributor.

**Comment by MichaReiser** [comment]
@dataCenter430 let's first finish your other contribution

**Comment by MichaReiser** [comment]
I think the ideal solution here is what Carl mentioned: A lint rule that identifies patch calls that point nowhere. Ideally, the IDE would also support rename in exactly these locations. 

Using `__name__` simply feels more awkward to write and using `__name__` also doesn't seem to be something that the pytest community recommends (I couldn't find a mention in their docs).

What's not clear to me (I'm neither a pytest nor unittest user) is how common it is to patch nested objects:

```py
monkeypatch.setattr("mymodule.sub.attr", 42)
with patch("mymodule.sub.attr", 42):
    ...
```

The proposed fix doesn't t work for those instances; This could confuse users about when to pass a string vs. when to use `__name__`. I'd also argue that `__name__` is more difficult to understand than just the string path.

**Comment by tjkuson** [comment]
I do a lot of work in a codebase that follows the `__name__` convention, so would, in theory, value a rule that enforced it. However, it would be tricky to implement correctly, as something like this can't be fixed simply by using `__name__`.

```python
class Foo:
    bar = True


def test_foo(monkeypatch: pytest.MonkeyPatch) -> None:
    assert Foo.bar is True
    monkeypatch.setattr(Foo, "bar", False)
    assert Foo.bar is False
```

**Comment by dataCenter430** [comment]
Good point,  `__name__` only works for callables, not plain attributes like `Foo.bar.` That's a real gap in the proposed fix. 
I think the rule would need to either drop the `__name__` suggestion entirely and just flag the pattern, or limit itself to cases where the patched attribute is known to be a callable. The latter needs type information though.
